lib/Conversion/TorchToTcp/TcpCustomOp.cpp

```diff
@@ -0,0 +1,111 @@
+//===------------------------------------------------------------*- C++ -*-===//
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// Also available under a BSD-style license. See LICENSE.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir-tcp/Conversion/TorchToTcp/TorchToTcp.h"
+
+#include "mlir-tcp/Dialect/IR/TcpDialect.h"
+#include "mlir-tcp/Dialect/IR/TcpOps.h"
+
+#include "PopulatePatterns.h"
+#include "Utils.h"
+#include "torch-mlir/Conversion/Utils/Utils.h"
+#include "torch-mlir/Dialect/Torch/IR/TorchOps.h"
+#include "torch-mlir/Dialect/Torch/Utils/Utils.h"
+
+#include "llvm/ADT/StringSet.h"
+
+using namespace mlir;
+using namespace mlir::tcp;
+using namespace mlir::torch;
+using namespace mlir::torch::Torch;
+
+namespace {
+class ConvertAtenGatherOp : public OpConversionPattern<AtenGatherOp> {
```

Comment:
What if we add a `tcp.gather` op because we want to codegen that? Do you plan to provide some flag to control this, for existing Tcp ops?

IMO we should not have custom op transformations for ops that potentially have in TCP.

Reply:
There are a few ways to support this: 
1. Explicit: Support the two paths in ConvertAtenGatherOp and control that from a pass option
1. Explicit: Keep them as two separate conversion patterns and control that from frontend pipeline (phase ordering + selective conversions)
1. Implicit: Lean towards codegen first (when available) and fallback to tcp.custom_op if not

Reply:
To elaborate a bit more, here are the options to support this use-case:

**Option 1:**
Add a bool flag (say `enableCustomOpPatternsInstead`) to `ConvertTorchToTcp` constructor ([here](https://sourcegraph.com/github.com/cruise-automation/mlir-tcp/-/blob/lib/Conversion/TorchToTcp/TorchToTcp.cpp)), and selectively include either tcp op patterns or tcp custom op patterns like so:
```
if (enableCustomOpPatternsInstead) {
    torch_to_tcp::populateTcpCustomOpPatternsAndLegality(
        typeConverter, patterns, target, convertTorchOpsSet);
} else {
    torch_to_tcp::populateElementwisePatternsAndLegality(
        typeConverter, patterns, target, convertTorchOpsSet);

    torch_to_tcp::populateMiscPatternsAndLegality(typeConverter, patterns,
                                                  target, convertTorchOpsSet);

    torch_to_tcp::populateDataMovementPatternsAndLegality(
        typeConverter, patterns, target, convertTorchOpsSet);
}
```

**Option 2:**
Remove `torch_to_tcp::populateTcpCustomOpPatternsAndLegality` from `ConvertTorchToTcp` entirely, and create a separate `ConvertTorchToTcpCustomOps` conversion pass instead. We can then include `ConvertTorchToTcpCustomOps` after `ConvertTorchToTcp` in our pipeline (to prefer codegen over kernels) but also craft all combinations to override this easily.

To me both these options look pretty similar, but I am leaning towards 2 over 1 because:

- organizationally I like 2 more than 1 because it reduces the mental overhead on the pipeline stitching side a bit
- 1 is slightly non-idiomatic because we have to be careful about avoiding to populate two patterns that match the same torch source op (e.g. aten.gather). Keeping them as a separate pass prevents that from happening by design.
- 2 allows future extensions such as generic custom op conversion more easily

PLMK your thoughts, or if you have other ideas. I can send a PR for either options relatively quickly, but I wanted to check-in with you first.

Reply:
Decided to go with Option 2 for now: https://github.com/cruise-automation/mlir-tcp/pull/19